awx/main/models/events.py

```diff
@@ -0,0 +1,774 @@
+        if artifact_dict:
+            if event_data and isinstance(event_data, dict):
+                # Note: Core has not added support for marking artifacts as
+                # sensitive yet. Going forward, core will not use
+                # _ansible_no_log to denote sensitive set_stats calls.
+                # Instead, they plan to add a flag outside of the traditional
+                # no_log mechanism. no_log will not work for this feature,
+                # in core, because sensitive data is scrubbed before sending
+                # data to the callback. The playbook_on_stats is the callback
+                # in which the set_stats data is used.
+
+                # Again, the sensitive artifact feature has not yet landed in
+                # core. The below is how we mark artifacts payload as
+                # senstive
+                # artifact_dict['_ansible_no_log'] = True
+                #
+                parent_job = self.objects.filter(pk=pk).first()
+                if hasattr(parent_job, 'artifacts') and parent_job.artifacts != artifact_dict:
+                    parent_job.artifacts = artifact_dict
+                    parent_job.save(update_fields=['artifacts'])
+
+        return job_event
+
+    @property
+    def job_verbosity(self):
+        return 0
+
+
+class JobEvent(BasePlaybookEvent):
+    '''
+    An event/message logged from the callback when running a job.
+    '''
+
+    VALID_KEYS = BasePlaybookEvent.VALID_KEYS + ['job_id']
+
+    class Meta:
+        app_label = 'main'
+        ordering = ('pk',)
+        index_together = [
+            ('job', 'event'),
+            ('job', 'uuid'),
+            ('job', 'start_line'),
+            ('job', 'end_line'),
+            ('job', 'parent_uuid'),
+        ]
+
+    job = models.ForeignKey(
+        'Job',
+        related_name='job_events',
+        on_delete=models.CASCADE,
+        editable=False,
+    )
+    host = models.ForeignKey(
+        'Host',
+        related_name='job_events_as_primary_host',
+        null=True,
+        default=None,
+        on_delete=models.SET_NULL,
+        editable=False,
+    )
+    host_name = models.CharField(
+        max_length=1024,
+        default='',
+        editable=False,
+    )
+    hosts = models.ManyToManyField(
+        'Host',
+        related_name='job_events',
+        editable=False,
+    )
+    parent = models.ForeignKey(
+        'self',
+        related_name='children',
+        null=True,
+        default=None,
+        on_delete=models.SET_NULL,
+        editable=False,
+    )
+    parent_uuid = models.CharField(
+        max_length=1024,
+        default='',
+        editable=False,
+    )
+
+    def get_absolute_url(self, request=None):
+        return reverse('api:job_event_detail', kwargs={'pk': self.pk}, request=request)
+
+    def __unicode__(self):
+        return u'%s @ %s' % (self.get_event_display2(), self.created.isoformat())
+
+    def _update_from_event_data(self):
```

Comment:
Project updates don't really have nested events, and they don't have a host reference (it's just `-i localhost,`, so these methods are specific to end-user playbook runs.